### DISCUSSION TITLE
None of the following functions can be called with the arguments supplied.

### DISCUSSION THREAD

**Discussion by ItSNeverLate** [jwtk/jjwt#977] [Q&A] [answered]
Hello everyone, I could not use this fun signature! Any idea?

.signWith(privateKey, Jwts.SIG.RS512)

` val kf = KeyFactory.getInstance("RSA")
        val pkcs8EncodedKeySpec = PKCS8EncodedKeySpec(accessKeyBase64.decodeBase64Bytes())
        val privateKey: Key = kf.generatePrivate(pkcs8EncodedKeySpec)
        val calendar = Calendar.getInstance()

        return Jwts.builder()
            .claims(payload)
            .issuedAt(calendar.time)
            .expiration(calendar.apply { add(Calendar.SECOND, expiresInSeconds) }.time)
            .header().add("typ", "JWT").add(headers)
            .and()
            // TODO: .signWith(privateKey) -> RSA256? should be checked by backend
            .signWith(privateKey, SignatureAlgorithm.RS512)
            .compact()`

other signatures work, but:
.signWith(privateKey, SignatureAlgorithm.RS512) --> is deprecated
.signWith(privateKey) --> uses RS256 while I need RS512

None of the following functions can be called with the arguments supplied.
signWith(TypeVariable(K)!, SecureDigestAlgorithm<in TypeVariable(K)!, *>!)   where K = TypeVariable(K) for    fun <K : Key!> signWith(key: K!, alg: SecureDigestAlgorithm<in K!, *>!): JwtBuilder! defined in io. jsonwebtoken. JwtBuilder
signWith(SignatureAlgorithm!, Key!) defined in io. jsonwebtoken. JwtBuilder
signWith(SignatureAlgorithm!, ByteArray!) defined in io. jsonwebtoken. JwtBuilder
signWith(SignatureAlgorithm!, String!) defined in io. jsonwebtoken. JwtBuilder
signWith(Key!, SignatureAlgorithm!) defined in io. jsonwebtoken. JwtBuilder

**Comment by lhazlewood** [comment, marked as answer]
Hi @ItSNeverLate !

```java
...
.signWith(privateKey, SignatureAlgorithm.RS512)
.compact()
```

should be:

```java
...
.signWith(privateKey, Jwts.SIG.RS512) // <-- Jwts.SIG.RS512
.compact()
```

The `.signWith` method signature requires that key specified must be compatible with the paired `SignatureAlgorithm` instance.  As such, the type of key required by `Jwts.SIG.RS512` must be a `PrivateKey`, but it looks like you've typed `val privateKey` to be just a `Key` (and not a `PrivateKey` as required).

I think that's the issue - let us know!

**Reply by ItSNeverLate** [reply]
You are right! Thank you for the help.
